crux_macros/src/effect.rs

```diff
@@ -199,6 +224,41 @@ mod tests {
                 }
             }
         }
+        #[cfg(feature = "typegen")]
+        impl ::crux_core::typegen::Export for Capabilities {
+            fn register_types(
+                generator: &mut ::crux_core::typegen::TypeGen,
+            ) -> ::crux_core::typegen::Result {
+                generator
+                    .register_type::<
+                        <Render<Event> as ::crux_core::capability::Capability<Event>>::Operation,
+                    >()?;
+                generator
+                    .register_type::<
+                        <<Render<
+                            Event,
+                        > as ::crux_core::capability::Capability<
+                            Event,
+                        >>::Operation as ::crux_core::capability::Operation>::Output,
+                    >()?;
+                generator.register_type::<EffectFfi>()?;
+                generator.register_type::<::crux_core::bridge::Request<EffectFfi>>()?;
+                Ok(())
+            }
+        }
+        #[cfg(not(feature = "typegen"))]
```

Comment:
This looks sus to me - this code is going to be in the _users_ crate right?  Are we expecting them to declare a `typegen` feature?

Reply:
(I've only skimmed the latest iteration of the code, so apologies if I missed something)

Reply:
Never mind, I see from the examples that we are.  Seems a bit odd though tbh.  I'd have thought users would either want typegen or not want typegen, rather than wanting to enable/disable it?

Reply:
Annoyingly, yes, because we need that feature in the macro generated code. Unless there's a way to make the macro itself conditional, which only just occured to me.

Reply:
> We need the macro to generate the typegen code when used from shared_types but not when used in the shared crate.

Why is that?

Reply:
I am still curious as to the why (maybe it invalidates my thinking), but :

Would a separate derive for `Export` or a parameter to the existing derive that controls whether it generates an `Export` not work?  If users are fine just having typegen everywhere they don't need to worry about feature flags, if they want to use feature flags they can `#cfg_attr` out the parameter/derive as appropriate. 

It is a tiny bit more code for users, but a macro that just doesn't work if your crate doesn't declare a particular feature flag seems a bit off to me...

Reply:
I like your idea to have a separate derive for `Export` (will give that a go) and to `#cfg_attr` it out. But still struggling to see how you'd avoid typegen code being deployed onto a phone without a feature.

Reply:
@obmarg I've split out the derive macro for implementing `Export`. Have a look and see if you think that's better (I do). The feature is still there but using `#[cfg_attr(feature = "typegen", derive(Export))]`, which I think is also better.

Reply:
Looks kind of like the GitHub outage was making us talk over each other yesterday 😂.  I definitely did not see your explanation prior to some of my comments.  But cool, I'll take a look

Reply:
Had a look - nice, definitely think it's an improvement.  Lets people optimise for this if they want, but doesn't _force_ people to.

> But still struggling to see how you'd avoid typegen code being deployed onto a phone without a feature.

If you don't call `Export` in your runtime code it'll get dropped well before it makes it to a phone, won't it?.  

Suppose there's a compile time argument for feature flags (why pay for compiling `Export` when compiling `shared` if it's not going to be used) but I'm not sure that's clear cut either: having a feature flag like this probably means you end up compiling `shared` (and maybe some of `crux`?) twice: once with the `build.rs` set of features and once without...

Reply:
Huh interesting. OK cool, thanks for your help with this. It's now, quite rightly, up to users to decide how they manage that, we just provide a derive macro to help. (I updated the book to explain this).